libcasr/src/lua.rs

```diff
@@ -0,0 +1,337 @@
+//! Lua module implements `ParseStacktrace`, `CrashLineExt` and `Severity` traits for Lua reports.
+use crate::error::*;
+use crate::execution_class::ExecutionClass;
+use crate::severity::Severity;
+use crate::stacktrace::{CrashLine, CrashLineExt, DebugInfo};
+use crate::stacktrace::{ParseStacktrace, Stacktrace, StacktraceEntry};
+
+use regex::Regex;
+
+/// Structure provides an interface for save parsing lua exception.
+#[derive(Clone, Debug)]
+pub struct LuaException {
+    message: String,
+}
+
+impl LuaException {
+    /// Create new `LuaException` instance from stream
+    pub fn new(stream: &str) -> Option<Self> {
+        let re = Regex::new(r#"\S+: .+\n\s*stack traceback:\s*\n(?:.*\n)*.+: .+"#).unwrap();
+        let mat = re.find(stream)?;
+        Some(LuaException {
+            message: mat.as_str().to_string(),
+        })
+    }
+    /// Extract stack trace from lua exception.
+    pub fn extract_stacktrace(&self) -> Result<Vec<String>> {
+        LuaStacktrace::extract_stacktrace(&self.message)
+    }
+    /// Transform lua exception into `Stacktrace` type.
+    pub fn parse_stacktrace(&self) -> Result<Stacktrace> {
+        LuaStacktrace::parse_stacktrace(&self.extract_stacktrace()?)
+    }
+    /// Get lua exception as a vector of lines.
+    pub fn lua_report(&self) -> Vec<String> {
+        self.message
+            .split('\n')
+            .map(|s| s.trim().to_string())
+            .collect()
+    }
+}
+
+/// Structure provides an interface for processing the stack trace.
+pub struct LuaStacktrace;
+
+impl ParseStacktrace for LuaStacktrace {
+    fn extract_stacktrace(stream: &str) -> Result<Vec<String>> {
+        let stacktrace = stream
+            .split('\n')
+            .map(|l| l.to_string())
+            .collect::<Vec<String>>();
+        let Some(first) = stacktrace
+            .iter()
+            .position(|line| line.trim().ends_with("stack traceback:"))
+        else {
+            return Err(Error::Casr(
+                "Couldn't find traceback in lua report".to_string(),
+            ));
+        };
+
+        let re = Regex::new(r#".+:(?:|\d+:) in .+"#).unwrap();
+        Ok(stacktrace[first..]
+            .iter()
+            .map(|s| s.trim().to_string())
```

Comment:
you trim stacktrace lines twice, you can do this when declaring the `stacktrace` variable above

Reply:
Done